include/exadg/operators/inverse_mass_operator.h

```diff
@@ -60,14 +62,50 @@ class InverseMassOperator
     this->matrix_free = &matrix_free_in;
     dof_index         = dof_index_in;
     quad_index        = quad_index_in;
+
+    cellwise_inverse_mass_not_available = not(matrix_free->get_dof_handler(dof_index)
```

Comment:
But cellwise inverse mass is only not available for H(div) velocity elements, correct? The pressure elements are not effected.

Reply:
yes. The implementation of inverse mass operator is generic and does not make a difference between velocity/pressure.

Reply:
What do you think about using the exact Assert as in dealii: https://github.com/necioglu/dealii/blob/c1ef6fab04853405c037b3eb759482b4d30e5aac/include/deal.II/matrix_free/evaluation_kernels.h#L5601-L5603

Reply:
I do not understand the `<=` in this assert, and there is no error message apart from "not implemented". I think we need our own ExaDG version here.

Reply:
The `<=` pertains to the fact that the implementation in deal.II detects different symmetries in the basis functions: https://github.com/necioglu/dealii/blob/c1ef6fab04853405c037b3eb759482b4d30e5aac/include/deal.II/matrix_free/shape_info.h#L60-L120 - such as collocation, support for even-odd decomposition, or additional shape functions besides the regular polynomial ones. In a sense `element_type` is not the best name, but it can be explained from the way it evolved.

Reply:
This is the only topic I have not addressed yet. I couldn't find a good looking solution for checking if the element is FE_DGQ, rather then:

```
matrix_free->get_dof_handler(dof_index).get_fe() == 
dealii::FE_DGQ<dim>(matrix_free->get_dof_handler(dof_index).get_fe().degree);
```

And additionally we have the check for `.all_reference_cells_are_hyper_cube()`. This is why the statement doesn't look good. Do we have an easier way to check if the element is FE_DGQ?

Reply:
You could use `dynamic_cast` instead of creating a new object and comparing. It will be successful or fail accordingly.

Reply:
> And additionally we have the check for .all_reference_cells_are_hyper_cube(). This is why the statement doesn't look good. Do we have an easier way to check if the element is FE_DGQ?

We don't need `FE_DGQ`, do we? In fact, all tensor-product DG elements work, and also systems of `FE_DGQ`. So the check is probably more complicated than that (on the positive side, you do not need `all_reference_cells_are_hyper_cube`):
```c++
  FiniteElement<dim> const &fe = matrix_free->get_dof_handler(dof_index).get_fe();
  if (((dim == 2 && fe.first_quad_index == 0) || (dim == 3 && fe.first_hex_index == 0))
      &&
      fe.base_element(0).dofs_per_cell == Utilities::pow(fe.degree + 1, dim))
```
The first check looks whether all unknowns are located on the interior of the cells (for which, unfortunately, I do not think there is a dimension-independent way to look up), whereas the second checks that we have a tensor product. Does that make sense to you? There might be simpler ways to achieve what I want (so feel free to improve on this), but I think that is the important features.

Reply:
I think I could simplify the first check with 
`fe.dofs_per_vertex == 0`
since it checks if we have DG elements, correct?